diff --git a/Completion/Unix/Type/_remote_files b/Completion/Unix/Type/_remote_files
index 93e1b7f43..4d4a7abbf 100644
--- a/Completion/Unix/Type/_remote_files
+++ b/Completion/Unix/Type/_remote_files
@@ -60,10 +60,7 @@ if zstyle -T ":completion:${curcontext}:files" remote-access; then
     dirprefix=${dir}/
   fi
 
-  if [[ -z $QIPREFIX ]]
-    then rempat="${dirprefix}${PREFIX%%[^./][^/]#}\*"
-    else rempat="${dirprefix}${(q)PREFIX%%[^./][^/]#}\*"
-  fi
+  rempat="${dirprefix}${(q)PREFIX%%[^./][^/]#}\*"
 
   # remote filenames
   remfiles=(${(M)${(f)"$(
@@ -91,8 +88,8 @@ if zstyle -T ":completion:${curcontext}:files" remote-access; then
     remdispfq=(${${remdispf%[*=|]}//(#b)(${~esc[2]})/\\$match[1]})
     remdispdq=(${${remdispd%/}//(#b)(${~esc[2]})/\\$match[1]})
   else
-    remdispfq=(${(q)remdispf%[*=|]})
-    remdispdq=(${(q)remdispd%/})
+    remdispfq=(${remdispf%[*=|]})
+    remdispdq=(${remdispd%/})
   fi

   local -a autoremove
